Add missing tracking to onboarding when changing steps and reduce their confusing properties

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Closes #2104
Closes #1467

This PR adds the missing tracking to extension onboarding (Merchant Center) onboarding and ads onboarding when changing steps, and it also adds new properties and marks confusing properties as deprecated to reduce confusion in the future.

For ads onboarding:
- Add `gla_setup_ads` tracking to record step changes.
- Update the `gla_setup_ads` tracking in `SetupAdsTopBar` to align with the new tracking properties.

For extension onboarding:
- Add `gla_setup_mc` tracking to record step changes when clicking on the stepper.
- Adjust the `gla_setup_mc` tracking to reduce confusion in the future and also align with the `gla_setup_ads` tracking properties.

### Detailed test instructions:

1. Enable tracking debug tool
   - Open the Console tab of the browser DevTool.
   - Change the filter level to "All levels".
   - Run `localStorage.setItem( 'debug', 'wc-admin:*' )` and refresh the page to make it work.
   - It should be able to view tracking via the Console tab. Ref:
      ![1](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/698f85d8-e890-4ab6-939b-738acee6098f)
2. Test extension onboarding
   1. Go to the Get Started page.
   2. Click on both "Start listing products" buttons to see if the new tracking properties are correct and the old properties stay the same.
   3. Click on the back button at the top-left of the onboarding flow to check the tracking properties.
   4. Go to the onboarding flow.
   5. Changing steps via the "Continue" buttons and the stepper to check the new tracking and properties.
3. Test ads onboarding
   1. Go to the ads onboarding flow.
   2. Changing steps via the "Continue" buttons and the stepper to check the new tracking and properties.
   3. Click on the back button at the top-left to check the tracking properties.

### Changelog entry

> Add - The missing tracking to onboarding when changing steps.
> Tweak - The properties of the gla_setup_mc and gla_setup_ads tracking events to reduce their confusion.

eason9487 committed:
Move the tracking doc of `gla_setup_ads` to js/src/utils/recordEvent.js

eason9487 committed:
Add tracking to the ads onboarding to record step changes.

eason9487 committed:
Update the 'gla_setup_ads' tracking in `SetupAdsTopBar` to align with the new tracking properties.

eason9487 committed:
Adjust the 'gla_setup_mc' tracking to reduce confusion in the future and also align with the 'gla_setup_ads' tracking properties.

Ref: https://github.com/woocommerce/google-listings-and-ads/issues/1467

eason9487 committed:
Add tracking to the onboarding to record step changes when clicking on the stepper.

eason9487 committed:
Update src/Tracking/README.md

puntope commented on js/src/get-started-page/get-started-with-video-card/index.js:
```diff
@@ -67,9 +67,12 @@ const GetStartedWithVideoCard = () => {
+						triggered_by: 'start-onboarding-button',
+						action: 'go-to-onboarding',
+						context: 'get-started-with-video',
+						// 'target' and 'trigger' were deprecated and can be removed after Q1 2024.
```
I see in other tracking actions added in this PR we don't use any more `trigger` and `target`. So, if they are not necessary, why not to remove them now?

eason9487 replied:
As it will take some time for a larger number of user-end extensions to be updated to the new version, for tracking events that have already recorded `trigger` and `target`, keeping them for a period is to reduce the degree of data interruption due to event property adjustments when analyzing tracking data. It also follows the suggestion in https://github.com/woocommerce/google-listings-and-ads/issues/2104#issuecomment-1727532869.

puntope commented on js/src/setup-ads/ads-stepper/index.js:
```diff
@@ -12,24 +13,43 @@ import SetupAccounts from './setup-accounts';
 	};
 
+	const continueStep = ( to ) => {
+		const from = step;
+
+		recordEvent( 'gla_setup_ads', {
+			triggered_by: `step${ from }-continue-button`,
+			action: `go-to-step${ to }`,
+		} );
```
I guess this is to avoid a race condition between the current `step` and the `step` after being updated by `setStep( to );`.

If is like this, I would add a comment clarifying so.

eason9487 replied:
The `from` variable is to pair with the `to` variable so that the semantic would be a bit more readable than:

```js
recordEvent( 'gla_setup_ads', {
	triggered_by: `step${ step }-continue-button`,
	action: `go-to-step${ to }`,
} );
```

puntope replied:
Oh, I see, It was confusing for me why we did that. As the step variable is the current step.

puntope commented on js/src/setup-ads/ads-stepper/index.js:
```diff
@@ -12,24 +13,43 @@ import SetupAccounts from './setup-accounts';
 		}
 	};
 
+	const continueStep = ( to ) => {
```
At least in the new created functions would be nice to add JSDOC comments

eason9487 replied:
Added in https://github.com/woocommerce/google-listings-and-ads/pull/2111/commits/465a8e24c9dc6cecc32de918c2952faf8e6a38e7.

puntope commented on js/src/setup-ads/ads-stepper/index.js:
```diff
@@ -12,24 +13,43 @@ import SetupAccounts from './setup-accounts';
 	const handleStepClick = ( value ) => {
 		if ( value < step ) {
+			recordEvent( 'gla_setup_ads', {
+				triggered_by: 'stepper-button',
```
Since we are adding `step1-continue-button `, `step2-continue-button `...etc I wonder if it would be better to add the origin step here. i.e `step1-stepper-button`

eason9487 replied:
Tweaked in https://github.com/woocommerce/google-listings-and-ads/pull/2111/commits/6ed73ac53ae0935ad0b3182f8113ea63770c972e.

I guess which one would be better will depend on the users of tracking data, the goal here is to reduce the confusion of the data itself.

The `stepN-continue-button` naming means the continue button in step N. But the stepper buttons exist in all steps, so I named it `stepper-stepN-button`.

puntope commented:
Would be nice to add some unit tests covering these events

puntope commented on js/src/utils/recordEvent.js:
```diff
@@ -76,9 +76,11 @@ export const recordTablePageEvent = ( context, page, direction ) => {
- * @property {string} target Button ID
- * @property {string} trigger Action (e.g. `click`)
- * @property {string} context Indicates which CTA is clicked
+ * @property {string} triggered_by Button name
```
```suggestion
 * @property {string} triggered_by Indicates which button triggered this event
```

puntope commented on js/src/utils/recordEvent.js:
```diff
@@ -107,6 +109,16 @@ export const recordTablePageEvent = ( context, page, direction ) => {
+ * Triggered on events during ads onboarding
+ *
+ * @event gla_setup_ads
+ * @property {string} triggered_by Button name
```
```suggestion
 * @property {string} triggered_by Indicates which button triggered this event
```

eason9487 committed:
Tweak the JSDoc descriptions of `triggered_by` for the 'gla_setup_mc' and 'gla_setup_ads' tracking.

Address:
- https://github.com/woocommerce/google-listings-and-ads/pull/2111#discussion_r1337437258
- https://github.com/woocommerce/google-listings-and-ads/pull/2111#discussion_r1337441163

Co-authored-by: Miguel Pérez Pellicer <5908855+puntope@users.noreply.github.com>

eason9487 committed:
Add JSDoc for `continueStep` functions.

Address: https://github.com/woocommerce/google-listings-and-ads/pull/2111#discussion_r1337385373

eason9487 committed:
Update src/Tracking/README.md

eason9487 committed:
Tweak the `triggered_by` tracking event property for the stepper in onboarding pages to indicate which stepper button is clicked.

Address: https://github.com/woocommerce/google-listings-and-ads/pull/2111#discussion_r1337431754

eason9487 commented:
Hi @puntope, thanks for the review!

This PR is ready for the second round of code reviews. Could you help with it?

puntope reviewed (approved):
✅  LGTM

Thanks for further adjustments.

mikkamp commented:
@eason9487 I was testing these changes and I noticed that for the regular Ads onboarding and if the ads account was connected during MC onboarding, then the stepper events work fine.

However when we create an additional campaign we also do this in a 2 step process:
1. Create paid campaign
2. Optimize your campaign

Going from step 1 > 2 we have a continue button without any tracking. Is that a completely different stepper within the code or can we get some similar tracking setup there? Let me know if I should create an additional issue.

eason9487 commented:
Hi @mikkamp

> However when we create an additional campaign we also do this in a 2 step process:
> 
> 1. Create paid campaign
> 2. Optimize your campaign
> 
> Going from step 1 > 2 we have a continue button without any tracking. Is that a completely different stepper within the code or can we get some similar tracking setup there?

Yes, these event tracking codes are not within the `<Stepper>` and `<AdsCampaign>` components, they are each different instances on each page.

> Let me know if I should create an additional issue.

As the event tracking to be added is similar, I think opening a PR and referencing this comment and #2104 would be fine. I will do so later.
